b2sdk/replication/monitoring.py

```diff
@@ -0,0 +1,281 @@
+######################################################################
+#
+# File: b2sdk/replication/monitoring.py
+#
+# Copyright 2022 Backblaze Inc. All Rights Reserved.
+#
+# License https://www.backblaze.com/using_b2_code.html
+#
+######################################################################
+
+import sys
+
+from abc import ABCMeta, abstractclassmethod, abstractmethod
+from collections import Counter
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass, field
+from queue import Queue
+from typing import ClassVar, Dict, Iterator, Optional, Tuple, Type
+
+from ..api import B2Api
+from ..bucket import Bucket
+from ..encryption.setting import EncryptionMode
+from ..file_lock import NO_RETENTION_FILE_SETTING, LegalHold
+from ..file_version import FileVersion
+from ..scan.folder import B2Folder
+from ..scan.path import B2Path
+from ..scan.policies import DEFAULT_SCAN_MANAGER, ScanPoliciesManager
+from ..scan.report import Report
+from ..scan.scan import zip_folders
+from .setting import ReplicationRule
+from .types import ReplicationStatus
+
+
+@dataclass(frozen=True)
+class AbstractReplicationAttrs(metaclass=ABCMeta):
+    """
+    Some attributes of files which are meaningful
+    for replication monitoring and troubleshooting.
+    """
+
+    @abstractclassmethod
+    def from_files(
+        cls, source_file: Optional[B2Path], destination_file: Optional[B2Path]
+    ) -> 'AbstractReplicationAttrs':
+        pass
+
+
+@dataclass(frozen=True)
+class ReplicationAttrs(AbstractReplicationAttrs):
+    """
+    Some attributes of source and destination files and their relations
+    which are meaningful for replication monitoring and troubleshooting.
+    """
+
+    # source attrs
+    source_replication_status: Optional[ReplicationStatus] = None
+    source_has_hide_marker: Optional[bool] = None
+    source_has_sse_c_enabled: Optional[bool] = None
+    source_has_large_metadata: Optional[bool] = None
+    source_has_file_retention: Optional[bool] = None
+    source_has_legal_hold: Optional[bool] = None
+
+    # destination attrs
+    destination_replication_status: Optional[ReplicationStatus] = None
+
+    # source & destination relation attrs
+    metadata_differs: Optional[bool] = None
```

Comment:
A pair of `(source path, destination path)` is used to populate to this structure. If source path is empty (because the file was deleted on source but not on destination), this will result in `source_...` fields being empty. Same for destination.